coveo-functools/README.md

```diff
@@ -4,116 +4,312 @@ Introspection, finalizers, delegates, dispatchers, waiters...
+```
 
-@flex
-class FlexibleConstructor:
-    def __init__(self, test: str) -> None:
-        self.test = test
+You could then feed it a payload like `{"name": "Lucy", address: [{"street": ...}, {...}]}`. 
+Flex will create an instance of `Person`, that has a list of 2 Address instances.
 
-    @flex
-    def flexible_method(self, test: str) -> str:
-        return test
+Note: The basic types `str, bool, int, float, dict, list, None` **are ignored** (no conversion occurs).
+This is because `json.load()` already returns these values in the proper type. This may change in the future.
 
 
-@flex
-def flexible_function(test: str) -> str:
-    return test
+### Supported objects and annotations
+
+Flex can be used with:
+- Classes and dataclasses
+- Functions
+- Methods
+- `Union[str, bool, int, float, list, dict, None]`  (or any combination of these basic **json-compatible types**)
+- These typing constructs, where `T` is your custom class:
+  - `List[T]`
+  - `Dict[str, T]`
+  - `Union[T, List[T]]`  (for APIs that may return a thing-or-list-of-things)
+  - `Optional[T]`
+
+
+### Limitations
+
+- Variable positional args (such as `def fn(*args): ...`) are left untouched.
+- Basic json-compatible types are left untouched. This is determined by the annotation, not the actual value.
+- If None is given as a value to deserialize into anything, None is given back. Absolutely no validation occurs in this case.
+- You can only `Union` basic json-compatible types, or `List[T], T`
+- No support for additional `typing` and `collections` objects other than the ones mentioned in this documentation.
+
+These are subject to change.
 
-instance = FlexibleConstructor(**PAYLOAD)
 
-assert instance.test == "SUCCESS"
-assert instance.flexible_method(**PAYLOAD) == "SUCCESS"
-assert flexible_function(**PAYLOAD) == "SUCCESS"
+## flex.deserialize
 
+This is where the magic happens, and is the recommended usage whenever it meets your use case. 
 
-# you can also use the tool inline; for instance to wrap a 3rd party lib:
-def typical_function(test: str) -> str:
-    return test
+TL;DR: Given that `payload` is a dict,`flex.deserialize(payload, hint=Job)` will convert `payload` into an instance of `Job`.
 
-assert flex(typical_function)(**PAYLOAD) == "SUCCESS"
+Here's an example puzzle! An uncanny API returns a messy "transaction" JSON:
+
+```json
+{
+    "Sold_To": {"Name": "Jon"},
+    "Items": [
+        {"sku": 123, "price": 19.99},
+        {"sku": 234, "price": 13.99},
+        {"sku": 0, "price": 0.50, "NOTE": "Forgot the reusable bag at home!!"}
+    ],
+    "Rebates": {
+        "airmiles": {"Flat": 10.0},
+        "coupon": [{"Flat": 0.79}, {"Flat":  1.50}],
+        "senior": {"Percentage": 2.5}
+    },
+    "Id": "GgfhAs89876yh.z"
+}
 ```
 
-Let's see a more practical example:
+
+Wouldn't it be convenient if you could create simple classes/dataclasses around them without any boilerplate?
+
+
+There are many  But you can solve it with flex. In one line, too!

```

Comment:
I think there's some leftover here. :smile:

Reply:
nah it was intentionally designed to catch your attention to my marketing pitch :trollface: